tests/test_transformations/test_knn_clustering.py

```diff
@@ -0,0 +1,205 @@
+        mean_10 = np.random.normal(loc=10, scale=1, size=(n_rows // 2, 2))
+        mean_100 = np.random.normal(loc=100, scale=1, size=(n_rows // 2, 2))
+        data_rows = np.vstack((mean_10, mean_100))
+
+        # Generate class labels, label 0 = mean 10, label 1 = mean 100
+        full_class_labels = np.array([0] * (n_rows // 2) + [1] * (n_rows // 2), dtype=object)
+        class_labels = np.array([x for x in full_class_labels],dtype=object)
+ 
+        # Replace ~50% of class labels with "missing" values
+        mask = np.random.rand(*full_class_labels.shape) < 0.5
+        class_labels[mask] = "no_label"
+
+        # annotation to test "missing_label" parameter
+        alt_class_labels = np.array([x for x in class_labels],dtype=object)
+        alt_class_labels[alt_class_labels == "no_label"] = np.nan
+        
+        # Combine data columns with class labels
+        self.syn_dataset = data_rows
+
+        self.syn_data = AnnData(
+            X=self.syn_dataset, var=pd.DataFrame(index=["gene1", "gene2"])
+        )
+
+        self.syn_data.layers["counts"] = self.syn_dataset
+
+        self.syn_data.obsm["derived_features"] = self.syn_dataset
+
+        self.syn_data.obs["classes"] = class_labels
+
+        # annotation with all labels missing
+        self.syn_data.obs["all_missing_classes"] = np.array(["no_label" for x in full_class_labels])
+        # annotation with all labels present
+        self.syn_data.obs["no_missing_classes"] = full_class_labels
+        self.syn_data.obs["alt_classes"] = alt_class_labels
+
+        # The string for column where class labels stored in obs
+        self.annotation = "classes"
+        self.alt_annotation = "alt_classes"
+        # The layer used for knn
+        self.layer = "counts"
+        # The features used for syn_data
+        self.syn_features = ["gene1", "gene2"]
+
+
+    def test_typical_case(self):
+        # This test checks if the function correctly adds 'knn' to the
+        # AnnData object's obs attribute and if it correctly sets
+        # 'knn_features' in the AnnData object's uns attribute.
+        knn_clustering(
+            adata=self.syn_data,
+            features=self.syn_features,
+            annotation=self.annotation,
+            layer=self.layer,
+        )
+        self.assertIn("knn", self.syn_data.obs)
+        self.assertEqual(self.syn_data.uns["knn_features"], self.syn_features)
+
+    def test_output_annotation(self):
+        # This test checks if the function correctly adds the "output_layer"
+        # to the # AnnData object's obs attribute
+        output_annotation_name = "my_output_annotation"
+        knn_clustering(
+            adata=self.syn_data,
+            features=self.syn_features,
+            annotation=self.annotation,
+            layer=self.layer,
+            output_annotation=output_annotation_name,
+        )
+        self.assertIn(output_annotation_name, self.syn_data.obs)
+
+    def test_layer_none_case(self):
```

Comment:
@oeldaghar, I suggest to put that test as part of _select_input_features if that is not already there.